drivers/adc/adc_mcux_12b1msps_sar.c

```diff
@@ -0,0 +1,308 @@
+}
+
+static void mcux_12b1msps_sar_adc_start_channel(const struct device *dev)
+{
+	const struct mcux_12b1msps_sar_adc_config *config = dev->config;
+	struct mcux_12b1msps_sar_adc_data *data = dev->data;
+
+	adc_channel_config_t channel_config;
+	uint32_t channel_group = 0U;
+
+	data->channel_id = find_lsb_set(data->channels) - 1;
+
+	LOG_DBG("Starting channel %d", data->channel_id);
+	channel_config.enableInterruptOnConversionCompleted = true;
+	channel_config.channelNumber = data->channel_id;
+	ADC_SetChannelConfig(config->base, channel_group, &channel_config);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct mcux_12b1msps_sar_adc_data *data =
+		CONTAINER_OF(ctx, struct mcux_12b1msps_sar_adc_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	mcux_12b1msps_sar_adc_start_channel(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct mcux_12b1msps_sar_adc_data *data =
+		CONTAINER_OF(ctx, struct mcux_12b1msps_sar_adc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void mcux_12b1msps_sar_adc_isr(const struct device *dev)
+{
+	const struct mcux_12b1msps_sar_adc_config *config = dev->config;
+	struct mcux_12b1msps_sar_adc_data *data = dev->data;
+	ADC_Type *base = config->base;
+	uint32_t channel_group = 0U;
+	uint16_t result;
+
+	result = ADC_GetChannelConversionValue(base, channel_group);
+	LOG_DBG("Finished channel %d. Result is 0x%04x", data->channel_id,
+		result);
+
+	*data->buffer++ = result;
+	data->channels &= ~BIT(data->channel_id);
+
+	if (data->channels) {
+		mcux_12b1msps_sar_adc_start_channel(dev);
+	} else {
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static int mcux_12b1msps_sar_adc_init(const struct device *dev)
+{
+	const struct mcux_12b1msps_sar_adc_config *config = dev->config;
+	struct mcux_12b1msps_sar_adc_data *data = dev->data;
+	ADC_Type *base = config->base;
+	adc_config_t adc_config;
+
+	ADC_GetDefaultConfig(&adc_config);
+
+	adc_config.referenceVoltageSource = config->ref_src;
+	adc_config.clockSource = config->clock_src;
+	adc_config.clockDriver = config->clock_drv;
+	adc_config.samplePeriodMode = config->sample_period_mode;
+	adc_config.resolution = kADC_Resolution12Bit;
+	adc_config.enableContinuousConversion = false;
+	adc_config.enableOverWrite = false;
+	adc_config.enableHighSpeed = false;
+	adc_config.enableLowPower = false;
+	adc_config.enableLongSample = false;
+	adc_config.enableAsynchronousClockOutput = true;
+
+	ADC_Init(base, &adc_config);
+
+#if !(defined(FSL_FEATURE_ADC_SUPPORT_HARDWARE_TRIGGER_REMOVE) && \
+	FSL_FEATURE_ADC_SUPPORT_HARDWARE_TRIGGER_REMOVE)
+	ADC_EnableHardwareTrigger(base, false);
+#endif
+
+	if (kStatus_Success == ADC_DoAutoCalibration(base)) {
+		LOG_DBG("ADC_DoAutoCalibration() Done.");
+	} else {
+		LOG_DBG("ADC_DoAutoCalibration() Failed.");
```

Comment:
```suggestion
		LOG_WRN("ADC_DoAutoCalibration() Failed.");
```

Reply:
hello, sorry for the long response, I add it to the last force update. I don't have enough time now to separate initial commit.